soc: lpcxpresso: Initial pinctrl implementation for NXP LPC

danieldegrasse opened the pull request:
Introduces initial pin group based pinctrl implementation for LPC SoCs. Currently, only the LPCXpresso55s69 EVK is enabled, although all drivers it uses support pinctrl

~Dependent on #44217 for CI failure fix in `clock_control_mcux_syscon.c`~

danieldegrasse committed:
dts: lpc55S6x: add pinctrl node to LPC SOC DTS

add soc level pinctrl node to LPC55s6x DTSI file, to be populated with
pin groups at the board level.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: add lpc pinctrl definition header

Add pinctrl definition header, to parse pinctrl groups in board level
DTS into pinctrl_soc_pin_t values.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pinctrl: Add LPC IOCON pinctrl driver

Add lpc iocon pinctrl driver. Driver handles IOCON clock initialization as
well as IOCON pin configuration

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: serial: Add pinctrl support to usart flexcomm driver

Add pinctrl support to LPC USART flexcomm driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: lpcxpresso55s69: Add pinctrl dts files for lpcxpresso55s69

Add pinctrl dtsi file for lpcxpresso55s69 board, as well as pinctrl
groups for lpcxpresso55s69 peripherals.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

gmarull commented on boards/arm/lpcxpresso55s69/lpcxpresso55s69-pinctrl.dtsi:
```diff
@@ -0,0 +1,94 @@
+		};
+	};
+
+	gpio1_default: gpio1_default {
+		group0 {
+			pinmux = <GPIO9_PIO1_9>,
+				<GPIO18_PIO1_18>;
+			bias-pull-up;
+			slew-rate = "standard";
+		};
+	};
```
note: I think this belongs to the GPIO domain, not pinctrl. Can you clarify this?

danieldegrasse replied:
That's a good point. I was under the impression that for GPIO devices, pinctrl drivers would handle the GPIO pin selection, and GPIO drivers themselves would handle setting pin state, such as input vs output.

In my mind the division was on a register level: the pinctrl driver would be responsible for writes to the pinmuxing `IOCON` peripheral, and the gpio driver would be responsible for writes to the `GPIO` peripheral. So there would still need to be a pinctrl group for the GPIO pins, to select them in the `IOCON` pinmux

danieldegrasse replied:
After looking at the GPIO driver for Kinetis I see what you mean. I've updated the LPC GPIO driver for pin control in this PR

gmarull replied:
note: STM32 has opened a PR to support static GPIO configuration via pinctrl, it may be useful for some edge cases, even though GPIO API is the way to go. I'll document that as part of the pinctrl guide. ref. https://github.com/zephyrproject-rtos/zephyr/pull/43961

gmarull reviewed (commented):
LGTM, I have one comment.

danieldegrasse committed:
drivers: gpio: gpio_mcux_lpc: Add pinmuxing control to gpio driver

Add pinmux control to gpio driver for LPC, so that pin control can work
correctly.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: i2c: i2c_mcux_flexcomm: enable pinctrl for i2c flexcomm driver

enable pinctrl for i2c flexcomm driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: spi: spi_mcux_flexcomm: add pinctrl support to lpc spi driver

add pinctrl support for spi_mcux_flexcomm driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pwm: pwm_mcux_sctimer: Add pinctrl support for lpc pwm

add pinctrl support for lpc sctimer pwm driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: i2s: i2s_mcux_flexcomm: add pinctrl support to lpc i2s driver

Add pinctrl support to i2s flexcomm driver for lpc.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: lpcxpresso55s69: remove pinmux settings

All peripherals used by lpcxpresso55s69 support pinctrl driver, so
remove pinmux settings. Pinmux file is retained for board specific I2S
loopback configuration.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@gmarull I've updated this PR to add pin control support for all drivers used by the LPC55s69, so that the pinmux settings can be removed (except for some board specific I2S loopback settings)

danieldegrasse committed:
tests: drivers: gpio: Remove pinmux setting on lpc for gpio_basic_api

LPC gpio driver now supports pinmux setting in gpio_pin_configure, so
remove the pinmux workaround for this platform in gpio_basic_api test.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbolivar-nordic commented on boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts:
```diff
@@ -43,15 +43,15 @@
 		user_button_1: button_0 {
 			label = "User SW1";
-			gpios = <&gpio0 5 GPIO_ACTIVE_HIGH>;
+			gpios = <&gpio0 5 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP)>;
```
Is this change really something for this patch?

danieldegrasse replied:
It's present because the LPCxpresso55s69 board `pinmux.c` file previously was setting the GPIO buttons to pull up in an initialization function. Without that pinmux call, these flags need to be added so that the gpio driver is aware these inputs should be pulled up.

mbolivar-nordic replied:
Ah, gotcha, thanks for the explanation. Resolved. Might want to add to the commit log but not a blocker.

mbolivar-nordic reviewed (commented):
Bindings LGTM. Nice work!

danieldegrasse committed:
boards: lpcxpresso55s69: Add pullup/pulldown fields to user buttons

Pinmux initialization function for lpcxpresso55s69 was setting user
button gpios as pullups, now that pin control will be used these
settings should be handled by the GPIO driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

gmarull reviewed (commented):
lgtm, one minor nitpick.

danieldegrasse commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -250,6 +256,14 @@ static int mcux_flexcomm_init(const struct device *dev)
+
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err) {
+		return -EINVAL;
```
Fixed, thanks

danieldegrasse committed:
dts: bindings: add lpc-iocon-pinctrl binding for LPC pinctrl

Add LPC pinctrl binding, for LPC IOCON pinctrl DTS node.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbolivar-nordic reviewed (approved):
gmarull-review-by-proxy
